File: models/experimental/deit/tt/deit_layer.py
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

from torch import nn
from typing import Optional, Tuple, Union

import ttnn

from tt_lib.fallback_ops import fallback_ops
from models.experimental.deit.tt.deit_config import DeiTConfig
from models.experimental.deit.tt.deit_attention import TtDeiTAttention
from models.experimental.deit.tt.deit_intermediate import TtDeiTIntermediate
from models.experimental.deit.tt.deit_output import TtDeiTOutput


class TtDeiTLayer(nn.Module):
    """This corresponds to the Block class in the timm implementation."""

    def __init__(self, config: DeiTConfig(), device, state_dict=None, base_address="") -> None:
        super().__init__()

        self.attention = TtDeiTAttention(config, device, state_dict, base_address=f"{base_address}.attention")
        self.intermediate = TtDeiTIntermediate(config, device, state_dict, base_address=f"{base_address}.intermediate")
        self.output = TtDeiTOutput(config, device, state_dict, base_address=f"{base_address}.output")

        ln_bw = state_dict[f"{base_address}.layernorm_before.weight"]
        ln_bb = state_dict[f"{base_address}.layernorm_before.bias"]
        self.layernorm_before = fallback_ops.LayerNorm(
            normalized_shape=config.hidden_size,
            weights=ln_bw,
            biases=ln_bb,
            eps=config.layer_norm_eps,
        )

        ln_aw = state_dict[f"{base_address}.layernorm_after.weight"]
        ln_ab = state_dict[f"{base_address}.layernorm_after.bias"]
        self.layernorm_after = fallback_ops.LayerNorm(
            normalized_shape=config.hidden_size,
            weights=ln_aw,
            biases=ln_ab,
            eps=config.layer_norm_eps,
        )

    def forward(
        self,
        hidden_states: ttnn.Tensor,
        head_mask: Optional[ttnn.Tensor] = None,
        output_attentions: bool = False,
    ) -> Union[Tuple[ttnn.Tensor, ttnn.Tensor], Tuple[ttnn.Tensor]]:
        self_attention_outputs = self.attention(
            self.layernorm_before(hidden_states),  # in DeiT, layernorm is applied before self-attention
            head_mask,
            output_attentions=output_attentions,
        )
        attention_output = self_attention_outputs[0]
        outputs = self_attention_outputs[1:]  # add self attentions if we output attention weights

        # first residual connection
        hidden_states = ttnn.add(attention_output, hidden_states)

        # in DeiT, layernorm is also applied after self-attention
        layer_output = self.layernorm_after(hidden_states)
        layer_output = self.intermediate(layer_output)

        # second residual connection is done here
        layer_output = self.output(layer_output, hidden_states)

        outputs = (layer_output,) + outputs

        return outputs

File: models/experimental/deit/deit_cpp/tt_cpp/deit_layer.cpp
// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0


#include "deit_layer.h"
#include "../helper_funcs.h"
#include <ttnn/operations/eltwise/binary/binary.hpp>
#include <ttnn/operations/normalization/layernorm/layernorm.hpp>
#include <ttnn/operations/data_movement/tilize/tilize.hpp>
#include <stdexcept>


TtDeiTLayer::TtDeiTLayer(
    const DeiTConfig& config,
    std::shared_ptr<ttnn::MeshDevice> device,
    const std::unordered_map<std::string, torch::Tensor>& state_dict,
    const std::string& base_address
) : config(config), device(device) {

    // Initialize sub-modules
    attention = std::make_unique<TtDeiTAttention>(
        config, device, state_dict, base_address + "attention."
    );
    intermediate = std::make_unique<TtDeiTIntermediate>(
        config, device, state_dict, base_address + "intermediate."
    );
    output = std::make_unique<TtDeiTOutput>(
        config, device, state_dict, base_address + "output."
    );

    // Load layer normalization parameters
    std::string ln_before_weight_key = base_address + "layernorm_before.weight";
    std::string ln_before_bias_key = base_address + "layernorm_before.bias";
    std::string ln_after_weight_key = base_address + "layernorm_after.weight";
    std::string ln_after_bias_key = base_address + "layernorm_after.bias";

    auto ln_before_weight_it = state_dict.find(ln_before_weight_key);
    auto ln_before_bias_it = state_dict.find(ln_before_bias_key);
    auto ln_after_weight_it = state_dict.find(ln_after_weight_key);
    auto ln_after_bias_it = state_dict.find(ln_after_bias_key);

    if (ln_before_weight_it == state_dict.end() || ln_before_bias_it == state_dict.end() ||
        ln_after_weight_it == state_dict.end() || ln_after_bias_it == state_dict.end()) {
        throw std::runtime_error("Layer normalization parameters not found in state_dict");
    }

    // Convert torch tensors to ttnn tensors
    layernorm_before_weight = helper_funcs::torch_to_tt_tensor_tile(ln_before_weight_it->second, device);
    layernorm_before_bias = helper_funcs::torch_to_tt_tensor_tile(ln_before_bias_it->second, device);
    layernorm_after_weight = helper_funcs::torch_to_tt_tensor_tile(ln_after_weight_it->second, device);
    layernorm_after_bias = helper_funcs::torch_to_tt_tensor_tile(ln_after_bias_it->second, device);
}

std::tuple<ttnn::Tensor, std::optional<ttnn::Tensor>> TtDeiTLayer::forward(
    const ttnn::Tensor& hidden_states,
    const std::optional<ttnn::Tensor>& head_mask,
    bool output_attentions
) {
    // Apply layer normalization before self-attention (pre-norm)
    auto normalized_hidden_states = helper_funcs::apply_layernorm(
        hidden_states,
        layernorm_before_weight,
        layernorm_before_bias,
        config.layer_norm_eps
    );


    // Self-attention
    auto attention_outputs = attention->forward(
        normalized_hidden_states,
        head_mask,
        output_attentions
    );

    auto attention_output = std::get<0>(attention_outputs);
    auto attention_weights = std::get<1>(attention_outputs);

    // First residual connection
    auto residual_output = ttnn::add(attention_output, hidden_states);

    // Apply layer normalization after self-attention
    auto layer_output = helper_funcs::apply_layernorm(
        residual_output,
        layernorm_after_weight,
        layernorm_after_bias,
        config.layer_norm_eps
    );


    // Intermediate layer (feed-forward)
    layer_output = intermediate->forward(layer_output);

    // Second residual connection (done in output layer)
    layer_output = output->forward(layer_output, residual_output);

    return std::make_tuple(layer_output, attention_weights);
}
